File: scripts/gen_resources.py
import os
import sys
import xml.etree.ElementTree as ET


def generate_resources(xml_path, header_path, source_path, init_fn="InitBinaryResources"):
    tree = ET.parse(xml_path)
    root = tree.getroot()
    base_dir = os.path.dirname(xml_path)

    # Generate Header
    with open(header_path, "w") as f:
        f.write("#pragma once\n\n")
        f.write("namespace AureliaUI {\n")
        f.write(f"    void {init_fn}();\n")
        f.write("} // namespace AureliaUI\n")

    # Generate Source
    with open(source_path, "w") as f:
        f.write(f'#include "{os.path.basename(header_path)}"\n')
        f.write("#include <core/ResourceManager.hpp>\n\n")
        f.write("namespace AureliaUI {\n\n")

        resources = []
        for file_tag in root.findall("file"):
            rel_path = file_tag.text.strip()
            abs_path = os.path.join(base_dir, rel_path)

            if not os.path.exists(abs_path):
                continue

            with open(abs_path, "rb") as res_file:
                content = res_file.read()

            var_name = "res_" + rel_path.replace("/", "_").replace(".", "_").replace(
                "-", "_"
            )
            f.write(f"static const uint8_t {var_name}[] = {{")
            f.write(",".join(hex(b) for b in content))
            f.write("};\n")
            resources.append((rel_path, var_name, len(content)))

        f.write(f"\nvoid {init_fn}() {{\n")
        f.write("    static bool initialized = false;\n")
        f.write("    if (initialized) return;\n")
        f.write("    initialized = true;\n\n")
        f.write("    auto& rm = ResourceManager::getInstance();\n")
        for rel_path, var_name, size in resources:
            f.write(
                f'    rm.registerResource("res://{rel_path}", {var_name}, {size});\n'
            )
        f.write("}\n\n")

        # Only the main kit uses static init (linked into Foundation). Per-app
        # bundles call their init from WinMain to avoid duplicate symbols.
        if init_fn == "InitBinaryResources":
            f.write("struct BinaryResourceInitializer {\n")
            f.write(f"    BinaryResourceInitializer() {{ {init_fn}(); }}\n")
            f.write("};\n\n")
            f.write("static BinaryResourceInitializer g_binaryResourceInitializer;\n\n")

        f.write("} // namespace AureliaUI\n")


if __name__ == "__main__":
    if len(sys.argv) < 4:
        print(
            "Usage: python gen_resources.py <resources.xml> <output.hpp> <output.cpp> "
            "[InitFunctionName]"
        )
        sys.exit(1)
    init_name = sys.argv[4] if len(sys.argv) >= 5 else "InitBinaryResources"
    generate_resources(sys.argv[1], sys.argv[2], sys.argv[3], init_name)

File: scripts/gen_resources.ts
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { basename, dirname, join } from "node:path";

function usage(): never {
  console.error(
    "Usage: bun run scripts/gen_resources.ts <resources.xml> <output.hpp> <output.cpp> [InitFunctionName]",
  );
  process.exit(1);
}

function decodeXmlText(value: string): string {
  return value
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
    .replace(/&amp;/g, "&");
}

function resourceFiles(xml: string): string[] {
  const files: string[] = [];
  const filePattern = /<file\b[^>]*>([\s\S]*?)<\/file>/gi;
  for (const match of xml.matchAll(filePattern)) {
    const file = decodeXmlText(match[1].trim());
    if (file.length > 0) files.push(file);
  }
  return files;
}

function variableName(path: string): string {
  return `res_${path.replace(/[\/\\.\-]/g, "_").replace(/[^A-Za-z0-9_]/g, "_")}`;
}

function bytesLiteral(bytes: Uint8Array): string {
  return Array.from(bytes, (byte) => `0x${byte.toString(16)}`).join(",");
}

function generateResources(xmlPath: string, headerPath: string, sourcePath: string, initFn: string): void {
  const xml = readFileSync(xmlPath, "utf8");
  const baseDir = dirname(xmlPath);
  const resources: Array<{ relPath: string; varName: string; size: number }> = [];

  mkdirSync(dirname(headerPath), { recursive: true });
  mkdirSync(dirname(sourcePath), { recursive: true });

  writeFileSync(
    headerPath,
    `#pragma once\n\nnamespace SphereUI {\n    void ${initFn}();\n} // namespace SphereUI\n`,
  );

  let source = `#include "${basename(headerPath)}"\n`;
  source += "#include <core/ResourceManager.hpp>\n\n";
  source += "namespace SphereUI {\n\n";

  for (const relPath of resourceFiles(xml)) {
    const absPath = join(baseDir, relPath);
    let content: Uint8Array;
    try {
      content = readFileSync(absPath);
    } catch {
      continue;
    }

    const varName = variableName(relPath);
    source += `static const uint8_t ${varName}[] = {${bytesLiteral(content)}};\n`;
    resources.push({ relPath, varName, size: content.length });
  }

  source += `\nvoid ${initFn}() {\n`;
  source += "    static bool initialized = false;\n";
  source += "    if (initialized) return;\n";
  source += "    initialized = true;\n\n";
  source += "    auto& rm = ResourceManager::getInstance();\n";
  for (const resource of resources) {
    source += `    rm.registerResource("res://${resource.relPath}", ${resource.varName}, ${resource.size});\n`;
  }
  source += "}\n\n";

  if (initFn === "InitBinaryResources") {
    source += "struct BinaryResourceInitializer {\n";
    source += `    BinaryResourceInitializer() { ${initFn}(); }\n`;
    source += "};\n\n";
    source += "static BinaryResourceInitializer g_binaryResourceInitializer;\n\n";
  }

  source += "} // namespace SphereUI\n";
  writeFileSync(sourcePath, source);
}

const [, , xmlPath, headerPath, sourcePath, initFn = "InitBinaryResources"] = Bun.argv;
if (!xmlPath || !headerPath || !sourcePath) usage();

generateResources(xmlPath, headerPath, sourcePath, initFn);
